Patching the Gaps: Essential Kibana Configurations added

muhammeddanish opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

### Description of the change

<!-- Describe the scope of your change - i.e. what the change does. -->

This pull request integrates five essential Kibana configurations into our Docker image, enhancing functionality and performance:
1. [server.publicBaseUrl](https://www.elastic.co/guide/en/kibana/8.11/settings.html#server-publicBaseUrl)
2. [xpack.security.encryptionKey](https://www.elastic.co/guide/en/kibana/current/using-kibana-with-security.html#security-configure-settings)
3. [xpack.reporting.encryptionKey](https://www.elastic.co/guide/en/kibana/8.11/reporting-settings-kb.html#encryption-keys)
4. [elasticsearch.requestTimeout](https://www.elastic.co/guide/en/kibana/8.11/settings.html#elasticsearch-requestTimeout)
5. [newsfeed.enabled](https://www.elastic.co/guide/en/kibana/8.11/settings.html)

### Benefits
<!-- What benefits will be realized by the code change? -->

This PR unlocks further Kibana customization possibilities by adding five essential configurations.
### Possible drawbacks
<!-- Describe any known limitations with your change -->

Without this update, Kibana in versions 7 and above won't work properly with custom domain names (DNS) due to the missing server.publicBaseUrl setting. Other changes enhance its overall functionality.

### Applicable issues
<!-- Enter any applicable Issues here (You can reference an issue using #) -->
By adding five key configurations, this PR opens doors to enhanced Kibana customization and fixes the issue of DNS access outlined in https://stackoverflow.com/questions/66634552/cant-access-kibana-with-domain.

DNS access for Kibana is now enabled by adding the server.publicBaseUrl configuration, a must for versions 7.x and above.

muhammeddanish committed:
adding few environment variiables

Signed-off-by: muhammeddanish <45929712+muhammeddanish@users.noreply.github.com>

carrodher commented:
Thanks for your contribution. Please note the version bump from `8.11.3` to `8.11.4` since this version is the upstream Kibana version which at this moment, the latest one is `8.11.3`, see https://github.com/elastic/kibana/releases.

Could you please revert any change related to this version bump?

muhammeddanish committed:
removing the version detais as recommended by github team

Signed-off-by: muhammeddanish <45929712+muhammeddanish@users.noreply.github.com>

muhammeddanish commented:
> Thanks for your contribution. Please note the version bump from `8.11.3` to `8.11.4` since this version is the upstream Kibana version which at this moment, the latest one is `8.11.3`, see https://github.com/elastic/kibana/releases.
> 
> Could you please revert any change related to this version bump?

Thanks @carrodher for the prompt response, I am not sure of how the versioning works in bitnami container images :) 

As recommended, I have removed the modified bump.

dgomezleon commented on bitnami/kibana/8/debian-11/rootfs/opt/bitnami/scripts/kibana-env.sh:
```diff
@@ -127,6 +132,16 @@ export KIBANA_FORCE_INITSCRIPTS="${KIBANA_FORCE_INITSCRIPTS:-no}"
 export SERVER_FORCE_INITSCRIPTS="$KIBANA_FORCE_INITSCRIPTS"
 export KIBANA_DISABLE_STRICT_CSP="${KIBANA_DISABLE_STRICT_CSP:-no}"
 export SERVER_DISABLE_STRICT_CSP="$KIBANA_DISABLE_STRICT_CSP"
+export KIBANA_SERVER_PUBLICBASEURL="${KIBANA_SERVER_PUBLICBASEURL:-}"
+export SERVER_PUBLICBASEURL="$KIBANA_SERVER_PUBLICBASEURL"
+export KIBANA_XPACK_SECURITY_ENCRYPTIONKEY="${KIBANA_XPACK_SECURITY_ENCRYPTIONKEY:-}"
+export XPACK_SECURITY_ENCRYPTIONKEY="$KIBANA_XPACK_SECURITY_ENCRYPTIONKEY"
+export KIBANA_XPACK_REPORTING_ENCRYPTIONKEY="${KIBANA_XPACK_REPORTING_ENCRYPTIONKEY:-}"
+export XPACK_REPORTING_ENCRYPTIONKEY="$KIBANA_XPACK_REPORTING_ENCRYPTIONKEY"
+export KIBANA_NEWSFEED_ENABLED="${KIBANA_NEWSFEED_ENABLED:-true}"
+export NEWSFEED_ENABLED="$KIBANA_NEWSFEED_ENABLED"
+export KIBANA_ELASTICSEARCH_REQUESTTIMEOUT="${KIBANA_ELASTICSEARCH_REQUESTTIMEOUT:-30000}"
+export ELASTICSEARCH_REQUESTTIMEOUT="$KIBANA_ELASTICSEARCH_REQUESTTIMEOUT"
```
From my understanding, these variables are only for Kibana, so you only need to add this:

```
export KIBANA_SERVER_PUBLICBASEURL="${KIBANA_SERVER_PUBLICBASEURL:-}"
export KIBANA_XPACK_SECURITY_ENCRYPTIONKEY="${KIBANA_XPACK_SECURITY_ENCRYPTIONKEY:-}"
export KIBANA_XPACK_REPORTING_ENCRYPTIONKEY="${KIBANA_XPACK_REPORTING_ENCRYPTIONKEY:-}"
export KIBANA_NEWSFEED_ENABLED="${KIBANA_NEWSFEED_ENABLED:-true}"
export KIBANA_ELASTICSEARCH_REQUESTTIMEOUT="${KIBANA_ELASTICSEARCH_REQUESTTIMEOUT:-30000}"
```

dgomezleon commented on bitnami/kibana/8/debian-11/rootfs/opt/bitnami/scripts/libkibana.sh:
```diff
@@ -215,6 +215,31 @@ kibana_initialize() {
             kibana_conf_set "csp.strict" "false" "bool"
         fi
 
+        #The publicly available URL that end-users access Kibana at
+        if ! is_empty_value "$SERVER_PUBLICBASEURL"; then
+            kibana_conf_set "server.publicBaseUrl" "$SERVER_PUBLICBASEURL"
+        fi
+
+        #Set an encryption key so that sessions are not invalidated. Use any text string that is 32 characters or longer as the encryption key. 
+        if ! is_empty_value "$XPACK_SECURITY_ENCRYPTIONKEY"; then
+            kibana_conf_set "xpack.security.encryptionKey" "$XPACK_SECURITY_ENCRYPTIONKEY"
+        fi
+        
+        #The static encryption key for reporting. Use an alphanumeric text string that is at least 32 characters. By default, Kibana generates a random key when it starts, which causes pending reports to fail 
+        if ! is_empty_value "$XPACK_REPORTING_ENCRYPTIONKEY"; then
+            kibana_conf_set "xpack.reporting.encryptionKey" "$XPACK_REPORTING_ENCRYPTIONKEY"
+        fi
+
+        #Controls whether to enable the newsfeed system for the Kibana UI notification center
+        if ! is_boolean_yes "$NEWSFEED_ENABLED"; then
+            kibana_conf_set "newsfeed.enabled" "false" "bool"
+        fi   
+
+        #Time in milliseconds to wait for responses from the back end or Elasticsearch. This value must be a positive integer. Default: 30000
+        if [[ "$ELASTICSEARCH_REQUESTTIMEOUT" != "30000" ]]; then
+            kibana_conf_set "elasticsearch.requestTimeout" "$ELASTICSEARCH_REQUESTTIMEOUT"
+        fi                              
+
```
All this block applies only for Kibana, so I would suggest this change:

```
        # Kibana override configuration
        if [[ "$SERVER_FLAVOR" = "kibana" ]]; then
            if is_boolean_yes "$KIBANA_DISABLE_STRICT_CSP"; then
                kibana_conf_set "csp.strict" "false" "bool"
            fi
            if ! is_empty_value "$KIBANA_SERVER_PUBLICBASEURL"; then
                kibana_conf_set "server.publicBaseUrl" "$KIBANA_SERVER_PUBLICBASEURL"
            fi
            if ! is_empty_value "$KIBANA_XPACK_SECURITY_ENCRYPTIONKEY"; then
                kibana_conf_set "xpack.security.encryptionKey" "$KIBANA_XPACK_SECURITY_ENCRYPTIONKEY"
            fi
            if ! is_empty_value "$KIBANA_XPACK_REPORTING_ENCRYPTIONKEY"; then
                kibana_conf_set "xpack.reporting.encryptionKey" "$KIBANA_XPACK_REPORTING_ENCRYPTIONKEY"
            fi
            if ! is_boolean_yes "$KIBANA_NEWSFEED_ENABLED"; then
                kibana_conf_set "newsfeed.enabled" "false" "bool"
            fi
            if [[ "$KIBANA_ELASTICSEARCH_REQUESTTIMEOUT" != "30000" ]]; then
                kibana_conf_set "elasticsearch.requestTimeout" "$KIBANA_ELASTICSEARCH_REQUESTTIMEOUT"
            fi
        fi
```

dgomezleon commented:
Hi @muhammeddanish,

Thanks for contributing! 
I have added some comments to the PR. Also, please note this should also be applied to branch 7.

muhammeddanish committed:
refactoring the code as per the suggestion from bitnami team

Signed-off-by: muhammeddanish <45929712+muhammeddanish@users.noreply.github.com>

muhammeddanish committed:
adding the changes to version 7

Signed-off-by: muhammeddanish <45929712+muhammeddanish@users.noreply.github.com>

muhammeddanish commented:
> Hi @muhammeddanish,
> 
> Thanks for contributing! I have added some comments to the PR. Also, please note this should also be applied to branch 7.

Thanks @dgomezleon !!
I have updated the code with the suggested changes.

dgomezleon reviewed (approved):
LGTM

muhammeddanish commented:
> LGTM

Thank you :)  

Could you please help me to merge this PR?
